File: YanDong/SortList.c
#include<stdio.h>
#include<memory.h>
#include<stdlib.h>
struct ListNode {
	int val;
	struct ListNode *next;
};
struct ListNode *getMid(struct ListNode *head){
	//printf("%d\n",head->next->val);
	struct ListNode *first=head,*second=head->next;
	while(first->next!=NULL && second->next!=NULL && second->next->next!=NULL){
		first = first->next;
		second = second->next->next;
	}
	return first;
}
struct ListNode* merge(struct ListNode *left,struct ListNode *right){
	struct ListNode *head = (struct ListNode*)malloc(sizeof(struct ListNode));
	struct ListNode *store = head;
	struct ListNode *l = left,*r = right;
	while(l != NULL && r !=NULL){
		if(l->val < r->val){
			head->next = l;
			head = l;
			l = l->next;
		}else{
			head->next = r;
			head = r;
			r = r -> next;
		}
	}
	if(l!=NULL){
		head->next = l;
	}
	if(r!=NULL){
		head->next = r;
	}
	struct ListNode *ret = store->next;
	free(store);
	return ret;
}

struct ListNode* sortList(struct ListNode* head) {
	//the end condition
	if(head == NULL || head->next == NULL){
		return head;
	}
	struct ListNode *mid = getMid(head);
	struct ListNode *right = mid->next;
	mid->next = NULL;
	struct ListNode *first = sortList(head);
	struct ListNode *second = sortList(right);
	return merge(first,second);
}
struct ListNode* init(){
	struct ListNode* dummy =(struct ListNode*)malloc(sizeof(struct ListNode));
	struct ListNode* head = dummy;
	int i;
	for(i = 10;i>1;--i){
		struct ListNode* node =(struct ListNode*)malloc(sizeof(struct ListNode));
		node->val = i;
		node->next = NULL;
		head->next = node;
		head = node;
	}
	struct ListNode* ret = dummy->next;
	free(dummy);
	return ret;
};
void print(struct ListNode *head){
	struct ListNode *dummy = head;
	while(dummy!=NULL){
		printf("%d ",dummy->val);
		dummy = dummy->next;
	}
	printf("\n");
}
int main(){
	struct ListNode *list = init();
	print(list);
	struct ListNode *ret = sortList(list);
	print(ret);
	getchar();
	return 0;
}

File: TangHongyin/sort_list.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):

    def sortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None:
            return None
        p,i = head,0
        while p:
            i += 1
            p = p.next
        return self.mergeSort(head, i)[0]

    def mergeSort(self, head, k):
        if k == 1:
            next = head.next
            head.next = None
            return head, next
        left_head, next = self.mergeSort(head, k/2)
        right_head, next = self.mergeSort(next, k-(k/2))
        return self.merge(left_head,right_head), next

    def merge(self, h1, h2):
        dummy = ListNode(0)
        p = dummy
        while h1 and h2:
            if h1.val <= h2.val:
                p.next = h1
                p = p.next
                h1 = h1.next
            else:
                p.next = h2
                p = p.next
                h2 = h2.next
        if h1 is None and h2 is not None:
            p.next = h2
        elif h2 is None and h1 is not None:
            p.next = h1
        return dummy.next